src/v/datalake/coordinator/state_update.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/coordinator/state_update.h"
+
+#include "datalake/coordinator/state.h"
+#include "datalake/coordinator/translated_offset_range.h"
+
+#include <iterator>
+
+namespace datalake::coordinator {
+
+checked<add_files_update, stm_update_error> add_files_update::build(
+  const topics_state& state,
+  const model::topic_partition& tp,
+  chunked_vector<translated_offset_range> entries) {
+    add_files_update update{
+      .tp = tp,
+      .entries = std::move(entries),
+    };
+    if (!update.can_apply(state)) {
+        return stm_update_error::failed;
+    }
+    return update;
+}
+
+bool add_files_update::can_apply(const topics_state& state) {
```

Comment:
it would be great to add some logging around the error handling here

Reply:
Will do, though likely in a follow up if this is ready to merge soon